tests/test_optimizer_random_dag.py

```diff
@@ -80,7 +82,18 @@ def generate_random_dag(
                     accelerators={
                         candidate.accelerator_name: candidate.accelerator_count
                     })
+                requested_features = set()
+                if op.num_nodes > 1:
+                    requested_features.add(
+                        clouds.CloudImplementationFeatures.MULTI_NODE)
+                try:
+                    resources.cloud.check_features_are_supported(
+                        resources, requested_features)
+                except exceptions.NotSupportedError:
+                    continue
```

Comment:
Quick question - instead of continuing, can we set `op.num_nodes = 1` and then `pass`? That way we can still include clouds which don't support multi-node in this test. (I may be missing something here)

```
                except exceptions.NotSupportedError:
                    op.num_nodes = 1
```

Reply:
We randomly set the `op.num_nodes` at the beginning so it should have some task that has single node, but this is a good point. I now enforce some of the tasks to be single node.

Reply:
Seems like we could end up having no candidate resources if it chooses a cloud that does not support multiple-node too many times? Maybe change to

```python
while True:
    candidate = random.choice(ALL_INSTANCE_TYPE_INFOS)
```

instead of using a fixed-sized array?

Reply:
It seems we are already choosing from the whole list of `ALL_INSTANCE_TYPE_INFOS` by setting `k=len(ALL_INSTANCE_TYPE_INFOS)`. It should be fine with the current way?